src/background/telemetry/index.js

```diff
@@ -76,52 +63,56 @@ const getConf = async (storage) => {
     installRandom: storage.installRandom,
     setup_shown: options.onboarding.shown,
   };
-};
-
-let telemetry;
-let telemetryEnabled = false;
+}
 
-chrome.runtime.onMessage.addListener((msg) => {
-  if (telemetryEnabled && msg.action === 'telemetry') {
-    telemetry.ping(msg.event);
-  }
-});
+let metrics;
+const setup = asyncSetup([
+  (async () => {
+    const storage = await loadStorage();
+    const { version } = chrome.runtime.getManifest();
+
+    metrics = new Metrics({
+      METRICS_BASE_URL: debugMode
+        ? 'https://staging-d.ghostery.com'
+        : 'https://d.ghostery.com',
+      EXTENSION_VERSION: version,
+      getConf: () => getConf(storage),
+      log: console.log.bind(console, '[telemetry]'),
+      storage,
+      saveStorage: (metrics) => {
+        saveStorage(storage, metrics);
+      },
+    });
 
-(async () => {
-  const storage = await loadStorage();
-  const { version } = chrome.runtime.getManifest();
+    // Just installed condition
+    if (storage.install_all === 0) {
+      const utms = await metrics.detectUTMs();
+      await chrome.storage.local.set({ utms });
+    } else {
+      const { utms = {} } = await chrome.storage.local.get(['utms']);
+      metrics.setUTMs(utms);
+    }
 
-  telemetry = new Telemetry({
-    METRICS_BASE_URL: debugMode
-      ? 'https://staging-d.ghostery.com'
-      : 'https://d.ghostery.com',
-    EXTENSION_VERSION: version,
-    getConf: () => getConf(storage),
-    log,
-    storage,
-    saveStorage: (metrics) => {
-      saveStorage(storage, metrics);
-    },
-  });
+    metrics.setUninstallUrl();
+  })(),
+]);
 
-  const JUST_INSTALLED = storage.install_all === 0;
+let enabled = false;
+observer.addListener('terms', async function telemetry(terms) {
+  enabled = terms;
 
-  try {
-    await recordUTMs(telemetry, JUST_INSTALLED);
-  } catch (error) {
-    log('Telemetry recordUTMs() error', error);
+  if (terms) {
+    setup.pending && (await setup.pending);
+    metrics.ping('install');
```

Comment:
should this have a condition ?

Reply:
No, I checked and the Metrics has the same: https://github.com/ghostery/ghostery-extension/blob/3ef12a5fb5dfd67c3de48f0f8f2750009ebe7af1/src/background/telemetry/metrics.js#L560-L561

Reply:
well, maybe we can move it outside of the metric, now it looks like we want to send install ping on every startup - which is not true

Reply:
It looks, like it looks, after all, if you don't change metrics, it will be checked twice. And I would keep the guard of internal data of Metrics inside of metrics...

Reply:
previous version had a check, new one does not.